rfcs/2021-02-04-5802-event-metadata.md

```diff
@@ -0,0 +1,206 @@
+   number of ways they may be lost in transit.
+
+Once in place, this metadata may also enable additional features not yet
+envisioned. For example, the VRL compiler collects type information on
+fields, but this can be expensive for arrays. Transforms such as
+`add_fields` could build a field type cache to improve the performance
+of this feature.
+
+The introduction of `SmallVec` in the metadata structure is driven by
+the assumption that most events will have a single source throughout
+their lifetime. `SmallVec` is a data structure that can inline a (fixed)
+number of elements before allocating memory. By using this feature, this
+common case can avoid an extra allocation and improve data locality.
+
+## Prior Art
+
+N/A
+
+## Drawbacks
+
+- The event data is a critical component of Vector. Any changes to this
+  structure impacts virtually the entire system.
+- All transforms that combine or duplicate events will have extra
+  complications in handing the metadata. This particularly applies to
+  the Lua and WASM scripting transforms.
+- This metadata will increase the memory required by Vector for any
+  given event flow.
+- Recording, copying, and finalizing the metadata will increase the CPU
+  usage.
+
+## Alternatives
+
+The following alternative components have been considered and discarded:
+
+1. Reduce the event metadata to a UUID, and track the actual metadata in
+   a separate map (or set of maps, as appropriate). This has the
+   downside of requiring most of the work necessary to just attach the
+   metadata itself to the event, as well as introducing both the
+   overhead of map lookups and the possibility of causing memory leaks
+   due to map entries not being purged, without any real benefit.
+2. Use a dictionary (ie `HashMap`) of arbitrary values to hold the
+   metadata. While some of the metadata may end up being optional, this
+   adds overhead to all accesses, due to the hashing and conversion of
+   the resulting data, as well as more work for the allocator.
+3. Store each event source in an `Arc`. Since the event source is
+   read-only, there is no need for an additional `Mutex` layer, and this
+   can provide an easy performance boost when events are duplicated
+   between multiple transforms or sinks. However, this adds the overhead
+   of an extra allocation and reference counting for the `Arc`
+   box. Since common usage patterns do not duplicate events, the costs
+   outweigh the benefits on this one.
+
+## Outstanding Questions
+
+- What is the best way of uniquely identifying the source? Is a simple
+  name adequate (since no two components may have the same name) or is a
+  new unique identifier or serial number required to prevent confusion
+  when configuration is reloaded?
+- What data is needed to uniquely identify source messages in order to
+  acknowledge them? The above provides for a `String`, into which any
+  other data may be encoded, but this is likely less than ideal.
```

Comment:
Should we defer this question to the end-to-end ack'ing RFC? It's likely we'll need to derive an ID from the event metadata, but I don't want that to distract from the purpose of this RFC.

Reply:
Gladly. I was trying to cover the data required for that use case, but if it will be fleshed out in another RFC then I can just leave that as stubs.